fix copy.deepcopy on LinearPackedParams

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61877
* __->__ #64367

Summary:

This is the same thing as https://github.com/pytorch/pytorch/pull/56154
but for quantized linear. It fixes the behavior of `copy.deepcopy` on
these modules. Before this PR, copied instances of `LinearPackedParams`
were not properly initialized, and inspecting them raised errors of
missing `_modules`. After this PR, inspecting and using the copies
works.

Test Plan:

```
python test/test_quantization.py TestStaticQuantizedModule.test_linear_api
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30702667](https://our.internmc.facebook.com/intern/diff/D30702667)

vkuzo committed:
fix copy.deepcopy on LinearPackedParams

Summary:

This is the same thing as https://github.com/pytorch/pytorch/pull/56154
but for quantized linear. It fixes the behavior of `copy.deepcopy` on
these modules. Before this PR, copied instances of `LinearPackedParams`
were not properly initialized, and inspecting them raised errors of
missing `_modules`. After this PR, inspecting and using the copies
works.

Need to resolve one TODO before landing (it is shown inline).

Test Plan:

```
python test/test_quantization.py TestStaticQuantizedModule.test_linear_api
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -155,15 +154,21 @@ def _test_linear_api_impl(self, batch_size, in_features, out_features, use_bias,
         self.assertEqual(qlinear.zero_point, loaded_qlinear.zero_point)
-        # make sure loaded_qlinear has the same dir as qlinear since
-        # scripting the module will add __overloads__ to __dict__
         self.checkScriptable(loaded_qlinear, [[X_q]], check_save_load=True)
```
actually should we do a copy before scripting so that it does not change load_qlinear?

jerryzh168 replied:
I know we just added copy in this PR but that feels like the correct thing to do..

vkuzo replied:
thanks, good point, updated the code to do that

vkuzo committed:
Update on "fix copy.deepcopy on LinearPackedParams"


Summary:

This is the same thing as https://github.com/pytorch/pytorch/pull/56154
but for quantized linear. It fixes the behavior of `copy.deepcopy` on
these modules. Before this PR, copied instances of `LinearPackedParams`
were not properly initialized, and inspecting them raised errors of
missing `_modules`. After this PR, inspecting and using the copies
works.

Test Plan:

```
python test/test_quantization.py TestStaticQuantizedModule.test_linear_api
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30702667](https://our.internmc.facebook.com/intern/diff/D30702667)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30702667).

jerryzh168 commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -155,15 +154,16 @@ def _test_linear_api_impl(self, batch_size, in_features, out_features, use_bias,
-        self.checkScriptable(loaded_qlinear, [[X_q]], check_save_load=True)
+        # scripting will add __overloads__ to __dict__, which is why we script a copy
+        # to be able to do the check in the next line
+        self.checkScriptable(copy.deepcopy(loaded_qlinear), [[X_q]], check_save_load=True)
```
nit: should probably do this in `checkScriptable` itself so that it works for other cases as well

vkuzo replied:
sure, can do this in a future PR

jerryzh168 reviewed (approved):
Looks good!
